README.md
globre
======

A python glob-like regular expression generation library.


CHANGELOG.rst
=========
ChangeLog
=========


v0.1.5
======

* Added support for alternate hierarchy separators


v0.1.4
======

* Added support for TravisCI (by @jayvdb)
* Added support for Python 2.6 (by @jayvdb)


v0.1.3
======

* Removed `distribute` dependency (thanks jlec)


v0.1.2
======

* First tagged release


globre/test.py
# -*- coding: utf-8 -*-
#------------------------------------------------------------------------------
# file: $Id$
# auth: metagriffin <mg.github@uberdev.org>
# date: 2013/09/22
# copy: (C) Copyright 2013-EOT metagriffin -- see LICENSE.txt
#------------------------------------------------------------------------------
# This software is free software: you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This software is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see http://www.gnu.org/licenses/.
#------------------------------------------------------------------------------

import unittest, re

import globre

#------------------------------------------------------------------------------
class TestGlobre(unittest.TestCase):

  if not hasattr(unittest.TestCase, 'assertIsNone'):
    def assertIsNone(self, *args, **kwargs):
      """Basic version of unittest.assertIsNotNone for Python 2.6 unittest."""
      return self.assertEqual(None, *args, **kwargs)

  if not hasattr(unittest.TestCase, 'assertIsNotNone'):
    def assertIsNotNone(self, *args, **kwargs):
      """Basic version of unittest.assertIsNotNone for Python 2.6 unittest."""
      return self.assertNotEqual(None, *args, **kwargs)


  #----------------------------------------------------------------------------
  def test_tokenizer(self):
    self.assertEqual(
      list(globre.Tokenizer('/foo/bar').tokens()),
      [('literal', '/foo/bar', 0, 8)])
    self.assertEqual(
      list(globre.Tokenizer('/foo\/bar').tokens()),
      [('literal', '/foo/bar', 0, 9)])
    self.assertEqual(
      list(globre.Tokenizer('/foo/*/bar').tokens()),
      [('literal', '/foo/', 0, 5),
       ('multiple', '*', 5, 6),
       ('literal', '/bar', 6, 10),
       ])
    self.assertEqual(
      list(globre.Tokenizer('/foo/**/bar/*.txt').tokens()),
      [('literal', '/foo/', 0, 5),
       ('any', '**', 5, 7),
       ('literal', '/bar/', 7, 12),
       ('multiple', '*', 12, 13),
       ('literal', '.txt', 13, 17),
       ])
    self.assertEqual(
      list(globre.Tokenizer('/foo/??.txt').tokens()),
      [('literal', '/foo/', 0, 5),
       ('single', '?', 5, 6),
       ('single', '?', 6, 7),
       ('literal', '.txt', 7, 11),
       ])
    self.assertEqual(
      list(globre.Tokenizer('/foo/??.txt').tokens()),
      [('literal', '/foo/', 0, 5),
       ('single', '?', 5, 6),
       ('single', '?', 6, 7),
       ('literal', '.txt', 7, 11),
       ])
    self.assertEqual(
      list(globre.Tokenizer(r'/foo/?\?.txt').tokens()),
      [('literal', '/foo/', 0, 5),
       ('single', '?', 5, 6),
       ('literal', '?.txt', 6, 12),
       ])
    self.assertEqual(
      list(globre.Tokenizer(r'/foo-[a-z0-9].txt').tokens()),
      [('literal', '/foo-', 0, 5),
       ('range', 'a-z0-9', 5, 13),
       ('literal', '.txt', 13, 17),
       ])
    self.assertEqual(
      list(globre.Tokenizer(r'/foo-{\\D{2,4\}}.txt').tokens()),
      [('literal', '/foo-', 0, 5),
       ('regex', r'\D{2,4}', 5, 16),
       ('literal', '.txt', 16, 20),
       ])

  #----------------------------------------------------------------------------
  def test_iswild(self):
    self.assertTrue(globre.iswild('/foo/bar/**.ini'))
    self.assertTrue(globre.iswild('/foo/bar-[0-9].ini'))
    self.assertFalse(globre.iswild('/foo/bar/conf.ini'))
    self.assertFalse(globre.iswild(r'\/foo\/bar\/conf.ini'))

  #----------------------------------------------------------------------------
  def test_match(self):
    self.assertIsNotNone(globre.match('/foo/bar/**.ini', '/foo/bar/conf.ini'))
    self.assertIsNotNone(globre.match('/foo/bar/**.ini', '/foo/bar/a/b/c/conf.ini'))
    self.assertIsNone(globre.match('/foo/bar/**.ini', '/a/foo/bar/conf.ini'))
    self.assertIsNone(globre.match('/foo/bar/**.ini', '/foo/bar/conf.ini.txt'))
    self.assertIsNone(globre.match('/foo/bar/**.ini', '/foo/bar/conf.ini/zig.txt'))
    self.assertIsNone(globre.match('/foo/bar/**.ini', '/FOO/BAR/ZIP/CONF.INI'))
    self.assertIsNotNone(globre.match('/foo/bar/**.ini', '/FOO/BAR/ZIP/CONF.INI', flags=re.IGNORECASE))

  #----------------------------------------------------------------------------
  def test_search(self):
    self.assertIsNotNone(globre.search('/foo/bar/**.ini', '/foo/bar/conf.ini'))
    self.assertIsNotNone(globre.search('/foo/bar/**.ini', '/foo/bar/a/b/c/conf.ini'))
    self.assertIsNotNone(globre.search('/foo/bar/**.ini', '/a/foo/bar/conf.ini'))
    self.assertIsNotNone(globre.search('/foo/bar/**.ini', '/foo/bar/conf.ini.txt'))
    self.assertIsNotNone(globre.search('/foo/bar/**.ini', '/foo/bar/conf.ini/zig.txt'))
    self.assertIsNone(globre.search('/foo/bar/**.ini', '/foo/bar/conf.txt'))
    self.assertIsNone(globre.search('/foo/bar/**.ini', '/X/FOO/BAR/ZIP/CONF.INI'))
    self.assertIsNotNone(globre.search('/foo/bar/**.ini', '/X/FOO/BAR/ZIP/CONF.INI', flags=re.IGNORECASE))

  #----------------------------------------------------------------------------
  def test_compile(self):
    expr = globre.compile('/foo/bar/*.dir/**.ini', flags=0)
    self.assertEqual(expr.pattern, r'\/foo\/bar\/[^/]*?\.dir\/.*?\.ini')
    expr = globre.compile('/foo/bar-??-[a-z0-9].ini', flags=0)
    self.assertEqual(expr.pattern, r'\/foo\/bar\-[^/][^/]\-[a-z0-9]\.ini')

  #----------------------------------------------------------------------------
  def test_compile_exact(self):
    expr = globre.compile('/foo/bar/*.dir/**.ini', flags=globre.EXACT)
    self.assertEqual(expr.pattern, r'^\/foo\/bar\/[^/]*?\.dir\/.*?\.ini$')
    self.assertIsNotNone(expr.match('/foo/bar/a.dir/blue/conf.ini'))
    self.assertIsNotNone(expr.match('/foo/bar/a.dir/conf.ini'))
    self.assertIsNone(expr.match('/foo/bar/blue/a.dir/conf.ini'))
    self.assertIsNone(expr.match('/foo/bar/a.dir/conf.ini.x'))
    self.assertIsNone(expr.match('/x/foo/bar/a.dir/conf.ini'))

  #----------------------------------------------------------------------------
  def test_prefix(self):
    self.assertEqual(globre.compile('/foo/bar', split_prefix=True)[0], '/foo/bar')
    self.assertEqual(globre.compile('/foo/b**', split_prefix=True)[0], '/foo/b')
    self.assertEqual(globre.compile('??/foo/b**', split_prefix=True)[0], '')

  #----------------------------------------------------------------------------
  def test_complete(self):
    expr = globre.compile(r'/foo[0-9a-f]/*/bar\[??\]/{\\D{2,4\}}/**.txt')
    self.assertEqual(
      expr.pattern,
      r'\/foo[0-9a-f]\/[^/]*?\/bar\[[^/][^/]\]\/\D{2,4}\/.*?\.txt')
    self.assertIsNotNone(expr.match('/foo6/zog/bar[16]/abra/cadabra.txt'))

  #----------------------------------------------------------------------------
  def test_sep_compile(self):
    self.assertEqual(
      globre.compile('!foo!bar!*.dir!**.ini', sep='!').pattern,
      r'\!foo\!bar\![^\!]*?\.dir\!.*?\.ini')

  #----------------------------------------------------------------------------
  def test_sep_default(self):
    self.assertIsNotNone(globre.match('/foo/**.ini', '/foo/bar/conf.ini'))
    self.assertIsNone(globre.match('/foo/**.ini', '\\foo\\bar\\conf.ini'))
    self.assertIsNotNone(globre.match(':foo:**.ini', ':foo:bar:conf.ini'))
    self.assertIsNone(globre.match('/foo/*.ini', '/foo/bar/conf.ini'))
    self.assertIsNone(globre.match('/foo/*.ini', '\\foo\\bar\\conf.ini'))
    self.assertIsNotNone(globre.match(':foo:*.ini', ':foo:bar:conf.ini'))

  #----------------------------------------------------------------------------
  def test_sep_alternate(self):
    self.assertIsNotNone(globre.match('/foo/**.ini', '/foo/bar/conf.ini', sep='/'))
    self.assertIsNotNone(globre.match('\\\\foo\\\\**.ini', '\\foo\\bar\\conf.ini', sep='\\'))
    self.assertIsNotNone(globre.match(':foo:**.ini', ':foo:bar:conf.ini', sep=':'))
    self.assertIsNone(globre.match('/foo/*.ini', '/foo/bar/conf.ini', sep='/'))
    self.assertIsNone(globre.match('\\\\foo\\\\*.ini', '\\foo\\bar\\conf.ini', sep='\\'))
    self.assertIsNone(globre.match(':foo:*.ini', ':foo:bar:conf.ini', sep=':'))

  #----------------------------------------------------------------------------
  def test_sep_multi(self):
    self.assertIsNotNone(globre.match('/foo/**.ini', '/foo/bar/conf.ini', sep='/\\'))
    self.assertIsNotNone(globre.match('/foo/**.ini', '\\foo\\bar\\conf.ini', sep='/\\'))
    self.assertIsNotNone(globre.match('/foo/**.ini', '/foo\\bar/conf.ini', sep='/\\'))
    self.assertIsNotNone(globre.match('\\\\foo\\\\**.ini', '\\foo\\bar\\conf.ini', sep='/\\'))
    self.assertIsNotNone(globre.match('\\\\foo\\\\**.ini', '/foo/bar/conf.ini', sep='/\\'))
    self.assertIsNotNone(globre.match('\\\\foo/**.ini', '\\foo\\bar\\conf.ini', sep='/\\'))
    self.assertIsNotNone(globre.match('\\\\foo/**.ini', '/foo\\bar/conf.ini', sep='/\\'))
    self.assertIsNone(globre.match('/foo/*.ini', '/foo/bar/conf.ini', sep='/\\'))
    self.assertIsNone(globre.match('/foo/*.ini', '\\foo\\bar\\conf.ini', sep='/\\'))
    self.assertIsNone(globre.match('/foo/*.ini', '/foo\\bar/conf.ini', sep='/\\'))
    self.assertIsNone(globre.match('\\\\foo\\\\*.ini', '\\foo\\bar\\conf.ini', sep='/\\'))
    self.assertIsNone(globre.match('\\\\foo\\\\*.ini', '/foo/bar/conf.ini', sep='/\\'))
    self.assertIsNone(globre.match('\\\\foo/*.ini', '\\foo\\bar\\conf.ini', sep='/\\'))
    self.assertIsNone(globre.match('\\\\foo/*.ini', '/foo\\bar/conf.ini', sep='/\\'))


#------------------------------------------------------------------------------
# end of $Id$
# $ChangeLog$
#------------------------------------------------------------------------------


globlib/test.py
# -*- coding: utf-8 -*-
#------------------------------------------------------------------------------
# file: $Id$
# auth: metagriffin <mg.github@uberdev.org>
# date: 2013/09/22
# copy: (C) Copyright 2013-EOT metagriffin -- see LICENSE.txt
#------------------------------------------------------------------------------
# This software is free software: you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This software is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see http://www.gnu.org/licenses/.
#------------------------------------------------------------------------------

import unittest, re

import globlib

#------------------------------------------------------------------------------
class TestGloblib(unittest.TestCase):

  #----------------------------------------------------------------------------
  def test_tokenizer(self):
    self.assertEqual(
      list(globlib.Tokenizer('/foo/bar').tokens()),
      [('literal', '/foo/bar', 0, 8)])
    self.assertEqual(
      list(globlib.Tokenizer('/foo\/bar').tokens()),
      [('literal', '/foo/bar', 0, 9)])
    self.assertEqual(
      list(globlib.Tokenizer('/foo/*/bar').tokens()),
      [('literal', '/foo/', 0, 5),
       ('multiple', '*', 5, 6),
       ('literal', '/bar', 6, 10),
       ])
    self.assertEqual(
      list(globlib.Tokenizer('/foo/**/bar/*.txt').tokens()),
      [('literal', '/foo/', 0, 5),
       ('any', '**', 5, 7),
       ('literal', '/bar/', 7, 12),
       ('multiple', '*', 12, 13),
       ('literal', '.txt', 13, 17),
       ])
    self.assertEqual(
      list(globlib.Tokenizer('/foo/??.txt').tokens()),
      [('literal', '/foo/', 0, 5),
       ('single', '?', 5, 6),
       ('single', '?', 6, 7),
       ('literal', '.txt', 7, 11),
       ])
    self.assertEqual(
      list(globlib.Tokenizer('/foo/??.txt').tokens()),
      [('literal', '/foo/', 0, 5),
       ('single', '?', 5, 6),
       ('single', '?', 6, 7),
       ('literal', '.txt', 7, 11),
       ])
    self.assertEqual(
      list(globlib.Tokenizer(r'/foo/?\?.txt').tokens()),
      [('literal', '/foo/', 0, 5),
       ('single', '?', 5, 6),
       ('literal', '?.txt', 6, 12),
       ])
    self.assertEqual(
      list(globlib.Tokenizer(r'/foo-[a-z0-9].txt').tokens()),
      [('literal', '/foo-', 0, 5),
       ('range', 'a-z0-9', 5, 13),
       ('literal', '.txt', 13, 17),
       ])
    self.assertEqual(
      list(globlib.Tokenizer(r'/foo-{\\D{2,4\}}.txt').tokens()),
      [('literal', '/foo-', 0, 5),
       ('regex', r'\D{2,4}', 5, 16),
       ('literal', '.txt', 16, 20),
       ])

  #----------------------------------------------------------------------------
  def test_iswild(self):
    self.assertTrue(globlib.iswild('/foo/bar/**.ini'))
    self.assertTrue(globlib.iswild('/foo/bar-[0-9].ini'))
    self.assertFalse(globlib.iswild('/foo/bar/conf.ini'))
    self.assertFalse(globlib.iswild(r'\/foo\/bar\/conf.ini'))

  #----------------------------------------------------------------------------
  def test_match(self):
    self.assertIsNotNone(globlib.match('/foo/bar/**.ini', '/foo/bar/conf.ini'))
    self.assertIsNotNone(globlib.match('/foo/bar/**.ini', '/foo/bar/a/b/c/conf.ini'))
    self.assertIsNone(globlib.match('/foo/bar/**.ini', '/a/foo/bar/conf.ini'))
    self.assertIsNone(globlib.match('/foo/bar/**.ini', '/foo/bar/conf.ini.txt'))
    self.assertIsNone(globlib.match('/foo/bar/**.ini', '/foo/bar/conf.ini/zig.txt'))
    self.assertIsNone(globlib.match('/foo/bar/**.ini', '/FOO/BAR/ZIP/CONF.INI'))
    self.assertIsNotNone(globlib.match('/foo/bar/**.ini', '/FOO/BAR/ZIP/CONF.INI', flags=re.IGNORECASE))

  #----------------------------------------------------------------------------
  def test_search(self):
    self.assertIsNotNone(globlib.search('/foo/bar/**.ini', '/foo/bar/conf.ini'))
    self.assertIsNotNone(globlib.search('/foo/bar/**.ini', '/foo/bar/a/b/c/conf.ini'))
    self.assertIsNotNone(globlib.search('/foo/bar/**.ini', '/a/foo/bar/conf.ini'))
    self.assertIsNotNone(globlib.search('/foo/bar/**.ini', '/foo/bar/conf.ini.txt'))
    self.assertIsNotNone(globlib.search('/foo/bar/**.ini', '/foo/bar/conf.ini/zig.txt'))
    self.assertIsNone(globlib.search('/foo/bar/**.ini', '/foo/bar/conf.txt'))
    self.assertIsNone(globlib.search('/foo/bar/**.ini', '/X/FOO/BAR/ZIP/CONF.INI'))
    self.assertIsNotNone(globlib.search('/foo/bar/**.ini', '/X/FOO/BAR/ZIP/CONF.INI', flags=re.IGNORECASE))

  #----------------------------------------------------------------------------
  def test_compile(self):
    expr = globlib.compile('/foo/bar/*.dir/**.ini', flags=0)
    self.assertEqual(expr.pattern, r'\/foo\/bar\/[^/]*?\.dir\/.*?\.ini')
    expr = globlib.compile('/foo/bar-??-[a-z0-9].ini', flags=0)
    self.assertEqual(expr.pattern, r'\/foo\/bar\-[^/][^/]\-[a-z0-9]\.ini')

  #----------------------------------------------------------------------------
  def test_compile_exact(self):
    expr = globlib.compile('/foo/bar/*.dir/**.ini', flags=globlib.EXACT)
    self.assertEqual(expr.pattern, r'^\/foo\/bar\/[^/]*?\.dir\/.*?\.ini$')
    self.assertIsNotNone(expr.match('/foo/bar/a.dir/blue/conf.ini'))
    self.assertIsNotNone(expr.match('/foo/bar/a.dir/conf.ini'))
    self.assertIsNone(expr.match('/foo/bar/blue/a.dir/conf.ini'))
    self.assertIsNone(expr.match('/foo/bar/a.dir/conf.ini.x'))
    self.assertIsNone(expr.match('/x/foo/bar/a.dir/conf.ini'))

  #----------------------------------------------------------------------------
  def test_prefix(self):
    self.assertEqual(globlib.compile('/foo/bar', split_prefix=True)[0], '/foo/bar')
    self.assertEqual(globlib.compile('/foo/b**', split_prefix=True)[0], '/foo/b')
    self.assertEqual(globlib.compile('??/foo/b**', split_prefix=True)[0], '')

  #----------------------------------------------------------------------------
  def test_complete(self):
    expr = globlib.compile(r'/foo[0-9a-f]/*/bar\[??\]/{\\D{2,4\}}/**.txt')
    self.assertEqual(
      expr.pattern,
      r'\/foo[0-9a-f]\/[^/]*?\/bar\[[^/][^/]\]\/\D{2,4}\/.*?\.txt')
    self.assertIsNotNone(expr.match('/foo6/zog/bar[16]/abra/cadabra.txt'))

#------------------------------------------------------------------------------
# end of $Id$
#------------------------------------------------------------------------------


globre/__init__.py
# -*- coding: utf-8 -*-
#------------------------------------------------------------------------------
# file: $Id$
# auth: metagriffin <mg.github@uberdev.org>
# date: 2013/09/22
# copy: (C) Copyright 2013-EOT metagriffin -- see LICENSE.txt
#------------------------------------------------------------------------------
# This software is free software: you can redistribute it and/or
# modify it under the terms of the GNU General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.
#
# This software is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see http://www.gnu.org/licenses/.
#------------------------------------------------------------------------------

import re

# new flags that apply to globs only...
EXACT = 1 << 10
E     = EXACT

#------------------------------------------------------------------------------
class Tokenizer(object):
  LITERAL  = 'literal'    # abcdef...
  SINGLE   = 'single'     # ?
  MULTIPLE = 'multiple'   # *
  ANY      = 'any'        # **
  RANGE    = 'range'      # [...]
  REGEX    = 'regex'      # {...}
  submap   = {
    '[' : (']', RANGE),
    '{' : ('}', REGEX),
    }
  def __init__(self, source):
    self.source = source
    self.pos    = 0
  def tokens(self):
    '''
    Generates four-element tuples of: (type, value, start, end).
    '''
    for token in self._outer():
      yield token
  def _outer(self):
    start = self.pos
    value = ''
    while self.pos < len(self.source):
      cur = self.source[self.pos]
      if cur not in '\\?*[{':
        value += cur
        self.pos += 1
        continue
      if cur == '\\':
        self.pos += 1
        if self.pos >= len(self.source):
          raise ValueError('dangling backslash "\\" in glob: %s' % (self.source,))
        value += self.source[self.pos]
        self.pos += 1
        continue
      if len(value) > 0:
        yield (self.LITERAL, value, start, self.pos)
      self.pos += 1
      value = ''
      start = self.pos
      if cur == '?':
        yield (self.SINGLE, '?', start - 1, start)
        continue
      if cur == '*':
        if self.pos >= len(self.source) or self.source[self.pos] != '*':
          yield (self.MULTIPLE, '*', start - 1, start)
          continue
        yield (self.ANY, '**', start - 1, start + 1)
        self.pos += 1
        start = self.pos
        continue
      if cur in self.submap:
        spec = self.submap[cur]
        value = self._scan(spec[0])
        if len(value) > 0:
          yield (spec[1], value, start - 1, self.pos)
        value = ''
        start = self.pos
        continue
      raise ValueError('unexpected glob character "%s" in glob: %s'
                       % (cur, self.source))
    if len(value) > 0:
      yield (self.LITERAL, value, start, self.pos)
  def _scan(self, target):
    value = ''
    while self.pos < len(self.source):
      cur = self.source[self.pos]
      self.pos += 1
      if cur == '\\':
        if self.pos >= len(self.source):
          raise ValueError('dangling backslash "\\" in glob: %s' % (self.source,))
        value += self.source[self.pos]
        self.pos += 1
        continue
      if cur == target:
        return value
      value += cur
    raise ValueError('no terminating "%s" in glob: %s' % (target, self.source))

WILDCHARS       = '?*[{\\'
SPECIAL_CHARS   = '?*[]{}'

#------------------------------------------------------------------------------
def iswild(pattern):
  for token in Tokenizer(pattern).tokens():
    if token[0] != Tokenizer.LITERAL:
      return True
  return False

#------------------------------------------------------------------------------
def compile(pattern, flags=0, sep=None, split_prefix=False):
  '''
  Converts a glob-matching pattern (using Apache Cocoon style rules)
  to a regular expression, which basically means that the following
  characters have special meanings:

  * ``?``:     matches any single character excluding the separator character
  * ``*``:     matches zero or more characters excluding the separator character
  * ``**``:    matches zero or more characters including the separator character
  * ``\``:     escape character used to precede any of the others for a literal
  * ``[...]``: matches any character in the specified regex-style range
  * ``{...}``: inlines a regex expression

  :Parameters:

  sep : str; default: "/"

    The `sep` parameter specifies the hierarchical path component
    separator to use. By default, it uses the unix-style forward-slash
    separator (``"/"``), but can be overriden to be a sequence of
    alternative valid hierarchical path component separator characters.
    Note that although `sep` *could* be set to both forward- and back-
    slashes (i.e. ``"/\\"``) to, theoretically, support either unix- and
    windows-style path components, this has the significant flaw that
    then *both* characters can be used within the same path as
    separators.

  flags : int; default: 0

    The `flags` bit mask can contain all the standard `re` flags, in
    addition to the ``globre.EXACT`` flag. If EXACT is set, then the
    returned regex will include a leading '^' and trailing '$', meaning
    that the regex must match the entire string, from beginning to end.

  split_prefix : bool; default: false

    If `split_prefix` is truthy, the return value becomes a tuple with
    the first element set to any initial non-wildcarded string found in
    the pattern. The second element remains the regex object as before.
    For example, the pattern ``foo/**.ini`` would result in a tuple
    equivalent to ``('foo/', re.compile('foo/.*\\.ini'))``.
  '''

  prefix = None
  expr   = ''

  if sep is None:
    sep = '/'
  if not sep:
    TypeError('invalid parameter "sep" value: %r' % (sep,))
  if set(sep) & set(SPECIAL_CHARS):
    TypeError('parameter "sep" cannot contain any of %r' % (SPECIAL_CHARS,))
  if len(sep) == 1:
    literal = re.escape
  else:
    def make_literal(sep):
      sep = '[' + re.escape(sep) + ']'
      sepcre = re.compile(sep)
      def _literal(text):
        return sep.join(sepcre.split(text))
      return _literal
    literal = make_literal(sep)
  if sep != '/':
    sep = re.escape(sep)

  for token in Tokenizer(pattern).tokens():
    if split_prefix and expr == '':
      prefix = token[1] if token[0] == Tokenizer.LITERAL else ''
    if token[0] == Tokenizer.LITERAL:
      expr += literal(token[1])
    elif token[0] == Tokenizer.SINGLE:
      expr += '[^' + sep + ']'
    elif token[0] == Tokenizer.MULTIPLE:
      expr += '[^' + sep + ']*?'
    elif token[0] == Tokenizer.ANY:
      expr += '.*?'
    elif token[0] == Tokenizer.RANGE:
      expr += '[' + token[1] + ']'
    elif token[0] == Tokenizer.REGEX:
      expr += token[1]
    else:
      ValueError('unexpected token %r from globre.Tokenizer for glob: %s'
                 % (token, pattern))

  if flags & EXACT:
    if not expr.startswith('^'):
      expr = '^' + expr
    # todo: technically, the last "$" *could* be escaped and therefore
    #       an extra "$" would need to be added... but that is very unlikely.
    if not expr.endswith('$'):
      expr += '$'

  expr = re.compile(expr, flags=flags & ~ EXACT)

  if prefix is not None:
    return (prefix, expr)
  return expr

#------------------------------------------------------------------------------
def match(pattern, string, sep=None, flags=0):
  flags |= EXACT
  return compile(pattern, sep=sep, flags=flags).match(string)

#------------------------------------------------------------------------------
def search(pattern, string, sep=None, flags=0):
  flags &= ~ EXACT
  return compile(pattern, sep=sep, flags=flags).search(string)

#------------------------------------------------------------------------------
# end of $Id$
#------------------------------------------------------------------------------
